SharpGenTools.Sdk/Documentation/DocProviderExecutor.cs

```diff
@@ -0,0 +1,299 @@
+            var docItem = cacheEntry ?? await QueryDocumentationProvider();
+
+            if (docItem == null)
+                return null;
+
+            element.Id = docItem.ShortId;
+            element.Description = docItem.Summary;
+            element.Remarks = docItem.Remarks;
+            docItem.Names.Add(docName);
+
+            if (cacheEntry == null)
+                cache.Add(docItem);
+
+            if (element.IsEmpty)
+                return docItem;
+
+            if (documentInnerElements)
+                DocumentInnerElements(element.Items, docItem);
+
+            return docItem;
+
+            async Task<IDocItem?> QueryDocumentationProvider()
+            {
+                if (docProvider == null)
+                    return null;
+
+                Lazy<string> docProviderName = new(
+                    () =>
+                    {
+                        try
+                        {
+                            var friendlyName = docProvider.UserFriendlyName;
+                            return string.IsNullOrWhiteSpace(friendlyName) ? FullName() : friendlyName;
+                        }
+                        catch
+                        {
+                            return FullName();
+                        }
+
+                        string FullName()
+                        {
+                            var type = docProvider.GetType();
+                            var name = type.FullName;
+                            return string.IsNullOrEmpty(name) ? type.Name : name;
+                        }
+                    },
+                    LazyThreadSafetyMode.None
+                );
+
+                List<Exception> exceptions = new(3);
+                var (backoff, backoffIndex, nextDelay) = GenerateBackoff(TimeSpan.Zero);
+
+                try
+                {
+                    for (uint retry = 0; retry <= 5; retry++)
+                    {
+                        if (retry != 0)
+                        {
+                            TimeSpan delay;
+                            if (nextDelay.HasValue)
+                                delay = nextDelay.Value;
+                            else
+                            {
+                                if (backoffIndex == 5)
+                                {
+                                    context.Logger.Message("SharpGen internal invalid state on delay: backoffIndex == 5");
+                                    if (Debugger.IsAttached) Debugger.Break();
+                                }
+                                else if (backoffIndex > 5)
+                                {
+                                    context.Logger.Message(
+                                        $"SharpGen internal invalid state on delay: backoffIndex = {backoffIndex}");
+                                    if (Debugger.IsAttached) Debugger.Break();
+                                    backoffIndex = 0;
+                                }
```

Comment:
I kept getting `IndexOutOfBoundsException`, but couldn't catch it in the debugger. This horrendous block is to assist with debugging, in case it appears again (I stopped hitting it). To be fixed before 2.0 RTW.

Reply:
Would using First-Chance Exceptions for `IndexOutOfBoundsException` enable us to break at the exception throw without having to have this workaround? IIRC that exception isn't enabled for first-chance in VS by default.

Reply:
I use Rider for SharpGen work (but e.g. VS for SharpGen SdkTests, I'm not consistent 😄), AFAIK it doesn't support breaking on first chance exceptions. I simplified the hack code though.